### PR TITLE
Improve performance of invocation diffing

### PR DESCRIPTION
**Pull request by bduffany**
* Switched from [diff](https://www.npmjs.com/package/diff) to Google's [diff-match-patch](https://github.com/google/diff-match-patch) which has much better performance. (Other people have issues with performance as well; e.g. https://github.com/kpdecker/jsdiff/issues/225). Tried it on a few different test builds; some cases that would take 5-10 seconds to load or even cause the browser to hang are now loading just as fast as the invocation overview page.

* Now hiding very long changed regions to avoid a massive amount of DOM nodes being rendered for invocations that are completely different.

* Added a preprocessing filter for progress events.

### CODE REVIEW

**Comment by bduffany** [comment]
Just realized that the diff is displaying a little oddly in one of those screenshots. Debugging it now

**Comment by bduffany** [comment]
Fixed the rendering issues, with the side effect of simplifying the implementation. See updated screenshot

File: app/compare/compare.css [hunk 1]
```diff
@@ -171,19 +165,33 @@
...
   padding: 0;
   margin: 0;
 }
 
-.compare-invocations .diff-line.removed {
-  color: #f44336;
+.compare-invocations del.removed {
+  background: #ffcdd2;
+  border-radius: 2px;
+  text-decoration: line-through;
```
Comment (siggisim) [inline app/compare/compare.css:176]: can we not strike through? adds a lot of visual noise - and i think the red probably says enough

File: app/compare/compare.css [hunk 2]
```diff
@@ -171,19 +165,33 @@
...
+.compare-invocations del.removed > :first-child {
+  color: black;
+}
+
+.compare-invocations .collapsed.removed {
+  background: #ffebee;
 }
 
-.compare-invocations .diff-line.added {
-  color: #43a047;
```
Comment (siggisim) [inline app/compare/compare.css:184]: kinda liked the old text color styling better vs background color

File: app/compare/diff_chunk.tsx [hunk 3]
```diff
@@ -24,38 +29,80 @@ export default class DiffChunk extends React.Component<DiffChunkProps, DiffChunk
...
-      </div>
-    ));
-
-    if (nodes.length >= 10 && isUnchanged && !this.state.expanded) {
-      return [
-        nodes[0],
-        <OutlinedButton className="diff-line collapsed" onClick={this.onExpand.bind(this)}>
-          <div className="plus-minus-cell">
+  private renderAdded(text: string) {
+    return <ins className="added">{text}</ins>;
```
Comment (siggisim) [inline app/compare/diff_chunk.tsx:38]: can these just be plain old div's or spans? same with del below

File: app/compare/diff_chunk.tsx [hunk 4]
```diff
@@ -24,38 +29,80 @@ export default class DiffChunk extends React.Component<DiffChunkProps, DiffChunk
...
+  }
+
+  private renderRemoved(text: string) {
+    return (
+      <del className="removed">
+        <span>{text}</span>
```
Comment (siggisim) [inline app/compare/diff_chunk.tsx:44]: do we need a del and a span? can we just go with div or span?

File: app/compare/diff_chunk.tsx [hunk 5]
```diff
@@ -1,61 +1,113 @@
...
 
-type DiffChunkState = {
+type State = {
   expanded: boolean;
 };
 
-export default class DiffChunk extends React.Component<DiffChunkProps, DiffChunkState> {
-  state = { expanded: this.props.defaultExpanded };
+const MIN_COLLAPSED_UNCHANGED_REGION_SIZE = 16;
+const MIN_COLLAPSED_CHANGED_REGION_SIZE = 128;
```
Comment (siggisim) [inline app/compare/diff_chunk.tsx:15]: what does "size" mean here? is it number of lines?

size in my mind is typically {Width, Height}

File: app/compare/diff_chunk.tsx [hunk 6]
```diff
@@ -1,61 +1,113 @@
...
+    );
+  }
+
+  private renderUnchanged(text: string) {
+    return <>{text}</>;
+  }
+
+  private renderChunk(
+    renderText: (text: string) => React.ReactNode,
+    collapsedLabel: string,
```
Comment (siggisim) [inline app/compare/diff_chunk.tsx:55]: are collapsedLabel and expandButtonClass ever different?

File: app/compare/compare.css (hunk 1 above)
Comment (bduffany) [reply app/compare/compare.css:176]: I did that for accessibility reasons - people with certain types of colorblindness (Deuteranopia or Achromatopsia) wouldn't be able to tell the difference between added and removed.

Comment (bduffany) [reply app/compare/compare.css:176]: FWIW this styling was basically copied from diff-match-patch's `diff_prettyHTML` function so I bet people at Google have thought through the practical concerns about the styling

File: app/compare/diff_chunk.tsx (hunk 5 above)
Comment (bduffany) [reply app/compare/diff_chunk.tsx:15]: Makes sense; replaced "SIZE" with "NUM_LINES"

File: app/compare/diff_chunk.tsx (hunk 3 above)
Comment (bduffany) [reply app/compare/diff_chunk.tsx:38]: This is needed for people with blindness to use this feature -- screenreaders would announce "ins" as "added" and "del" as "removed".

I think the more accessibility we can bake in, the better -- some teams might have blind engineers and lack of accessibility may be a dealbreaker

File: app/compare/diff_chunk.tsx (hunk 4 above)
Comment (bduffany) [reply app/compare/diff_chunk.tsx:44]: Span is needed for having the strikethrough be a different color than the text; see comments above re. accessibility

File: app/compare/diff_chunk.tsx (hunk 6 above)
Comment (bduffany) [reply app/compare/diff_chunk.tsx:55]: They're effectively different data types (one is a class name and one is a UI message) so I figured they should be separate params. I didn't want someone to change the label and not realize that it's a class name, and then they change it to something that's an invalid as a class name. Also if we translate the UI at some point in the future it's a bug waiting to happen if these aren't separate.

File: app/compare/compare.css (hunk 1 above)
Comment (bduffany) [reply app/compare/compare.css:176]: We could use `+` / `-` as visual indicators for accessibility instead, but this diff library that we're using now doesn't seem to have a way to diff on a line-by-line basis.

Comment (siggisim) [reply app/compare/compare.css:176]: The Github style + / - indicators which we used to have look a lot cleaner.

If you look at the examples in that repo - it was definitely created by engineers, not designers (and it hasn't been updated for 2-3 years).
https://neil.fraser.name/software/diff_match_patch/demos/diff.html

The biggest source of visual noise is the strikethrough whitespace.
Screen Shot 2020-11-06 at 1 23 12 PM

I could probably live with single line stike-throughs, but multi-line are pretty overboard.

File: app/compare/diff_chunk.tsx (hunk 3 above)
Comment (siggisim) [reply app/compare/diff_chunk.tsx:38]: According to [mdn docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ins), most screenreaders don't announce ins/del elements.

My biggest worry here is different browsers styling these rarely used element types differently.

More importantly I think we need to think about accessibility holistically, and not just sprinkle it in some places. There are lots of places in our app that are not accessible to those who are red/green colorblind (from Trends to the build history grid and beyond), or those using screen-readers (we use divs/spans across the board - not semantic elements).

We'll need to go through and add aria attributes, and can even add a colorblind mode in settings to support those use cases. This is something we should definitely do once we have the team and the usage to support it, but for the time being having a product that isn't accessible and cluttered seems like the worst of both worlds.

Comment (bduffany) [reply app/compare/diff_chunk.tsx]: Thanks for the explanation -- changed to regular divs & spans, and I'll keep that philosophy on accessibility in mind when working on future components.

image

File: app/compare/compare.css (hunk 1 above)
Comment (bduffany) [reply app/compare/compare.css]: OK, I found a way to use the diff-match-patch library in line mode, and reverted back to the previous styling:

image

File: yarn-error.log [hunk 7]
```diff
@@ -0,0 +1,2072 @@
...
+  1.22.4
+
+Node version: 
+  14.7.0
+
+Platform: 
+  linux x64
+
+Trace: 
+  Error: https://registry.yarnpkg.com/@types%2fline-diff: Not found
```
Comment (siggisim) [inline yarn-error.log:17]: probably don't want to check this in

**Review by siggisim — APPROVED** [review]
Thanks again for dealing with my comments - this looks great!

File: yarn-error.log (hunk 7 above)
Comment (bduffany) [reply yarn-error.log]: Removed, thanks!

**Pull request merged**